refactor(agw): replace scapy in gw mac address

wolfseb opened the pull request:
Closes #14442

With @MoritzThomasHuebner 

## Summary

This PR replaces usages of scapy in pipelined with:
* socket and dpkt for IPv4 calls
* `ip neigh get` for IPv6 calls

For the parsing of the byte string returned from the raw socket it introduces a class `ParseSocketPacket`.

## Test Plan

Run pipelined sudo tests and [lte integration tests](https://github.com/wolfseb/magma/actions/runs/3529160746/jobs/5919954767)

## Additional Information

- [ ] This change is backwards-breaking

MoritzThomasHuebner committed:
refactor(agw): replace scapy with socket and dpkt

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>
Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

Commit:
feat(agw): add packet parser

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

crasu commented on lte/gateway/python/magma/pipelined/app/packet_parser.py:
```diff
@@ -0,0 +1,181 @@
+    def __str__(self):
+        return f"VLAN tag\n" \
+               f"  prio  : {self.prio}\n" \
+               f"  id    : {self.cfi}\n" \
```
cfi is not the id

crasu commented on lte/gateway/python/magma/pipelined/app/packet_parser.py:
```diff
@@ -0,0 +1,181 @@
+
+def _parse_vlan_tag(packet: bytes) -> Tuple[Dot1Q, bytes]:
+    prio = packet[0] >> 5
+    cfi = packet[0] & 0b00011111
```
the cfi is a 1 bit field only this does not match the mask see:
https://en.wikipedia.org/wiki/IEEE_802.1Q#cite_ref-9

wolfseb replied:
Ah ok, yes you are right, thank you. How does this look:
```
prio = packet[0] >> 5
dei = packet[0] & 0b00010000 >> 4
vlan = packet[0] & 0b00001111 | packet[1]
```

crasu commented on lte/gateway/python/magma/pipelined/bridge_util.py:
```diff
@@ -96,22 +96,6 @@ def get_ofport(interface_name):
-        Gets L2 mac address of interface
-        """
-        try:
-            fl = "/sys/class/net/%s/address" % interface_name
```
maybe we should have used this implementation instead of the module

MoritzThomasHuebner committed:
refactor(agw): Added `get_mac_by_ip4` and `get_iface_by_ip4`

- Capable of getting local network IPs
- Added vlan argument to `_send_packet_and_receive_response`
- Fixed paket parsing for received pakets without VLAN

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

Commit:
chore(agw): add type hints to functions in gw_mac_address

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
refactor(agw): replace scapy in _get_gw_mac_address_v6

replace the scapy function getmacbyip6 with a wrapper, get_mac_by_ip6,
to the cli tool ip neigh get

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
refactor(agw): replace scapy call in ebpf_manager

Replace calls to `getmacbyip` and `getmacbyip6` with functions defined
in gw_mac_address.py.

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
chore(agw): remove scapy dependency from bazel files

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
fix(agw): fix unit tests

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Add dpkt to setup.py

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Reordering functions

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

LKreutzer commented on lte/gateway/python/magma/mobilityd/BUILD.bazel:
```diff
@@ -108,3 +108,9 @@ py_library(
+
+py_library(
+    name = "mac",
+    srcs = ["mac.py"],
```
This source file is already provided in the `dhcp_desc` library, could we just use that instead of creating a new one? Or does it make sense to split them up in general?

MoritzThomasHuebner replied:
We changed this as you suggested.

crasu commented on lte/gateway/python/magma/pipelined/gw_mac_address.py:
```diff
@@ -29,68 +30,108 @@ def get_gw_mac_address(ip: IPAddress, vlan: str, non_nat_arp_egress_port: str) -
+def get_mac_by_ip6(gw_ip: str) -> str:
+    for iface, ip6 in get_ifaces_by_ip6(gw_ip):
+        res = subprocess.run(
+            ["ip", "neigh", "get", gw_ip, "dev", iface],
```
This does not work ip neigh will only look into the neighbour table after the address has already been resolved.

crasu replied:
You could ping the host with -c before but this is a terrible workaround.

crasu replied:
May be you need something like: `sudo ip neigh del <ip> dev <devicename>` to  clear the cache before the ping.

MoritzThomasHuebner replied:
We added the a `ping -c 1 {gw_ip}` for now. The delete is not necessary to discern whether the mac still exists. Performing a ping on a unreachable device will simply set the entry in the ip neigh table to `FAILED` and return a non-zero exit code.

MoritzThomasHuebner replied:
We also did some research, but we could not find a better workaround for now.

MoritzThomasHuebner committed:
refactor(agw): Added ping to ip neigh check

- Ping the IP to refresh the ip neighbor table

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Simplified Bazel files

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

LKreutzer reviewed (approved):
Bazel lgtm

crasu reviewed (changes requested):
The ndp issue needs to be resolved as this would break non nat mode

crasu reviewed (approved):
LGTM

vktng reviewed (approved):
LGTM
